src/ua/main.cpp

```diff
@@ -529,6 +543,39 @@ void printEnvironmentInfo() {
   }
 }
 
+// There is currently the possibility of a race condition if a chunk
+// upload timed-out.  It's possible that a second upload succeeds,
+// has the chunk marked as "complete" and then the first request makes
+// its way through the queue and marks the chunk as pending again.
+// Since we are just about to close the file, we'll check to see if any
+// chunks are marked as pending, and if so, we'll retry them.
+void check_for_complete_chunks(vector<File> &files) {
+  for (int currCheckNum=0; currCheckNum < NUM_CHUNK_CHECKS; ++currCheckNum){
+    while (!chunksFinished.empty()) {
+      Chunk *c = chunksFinished.consume();
+      if (!is_chunk_complete(c)) {
+          chunksToUpload.produce(c);
+      }
+    }
+    // All of the chunks were marked as complete, so let's exit and we
+    // should be safeish to close the file.
+    if(chunksToUpload.size() == 0)
```

Comment:
Before we loop around and check for pending chunks, shouldn't we wait for `chunksToUpload` to drain?

```
while chunksToUpload is not empty:
  sleep
then loop again and describe all files
etc.
```